File: internal/translate/translate_test.go
package translate

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"

	"github.com/dat267/dscli/internal/deepseek"
)

// TestChunkTextHardSplitsLongLines: a single line longer than the chunk size
// is hard-split so a file with no newlines still chunks (the old behaviour
// left it as one oversized request whose output got cut off).
func TestChunkTextHardSplitsLongLines(t *testing.T) {
	const budget = 16 * 1024
	chunks := ChunkText(strings.Repeat("a", 90*1024), budget)
	want := (90*1024 + budget - 1) / budget
	if len(chunks) != want {
		t.Fatalf("chunks = %d, want %d", len(chunks), want)
	}
	if len(chunks[0]) != budget {
		t.Errorf("first chunk len = %d, want %d", len(chunks[0]), budget)
	}
	joined := strings.Join(chunks, "")
	if len(joined) != 90*1024 {
		t.Errorf("reassembled len = %d, want %d", len(joined), 90*1024)
	}
}

func TestResolveStyle(t *testing.T) {
	t.Run("default fallback", func(t *testing.T) {
		style, err := ResolveStyle("", "ja", "en")
		if err != nil {
			t.Fatalf("ResolveStyle: %v", err)
		}
		for _, want := range []string{"STYLE: GENERAL", "pronouns", "active voice", "register"} {
			if !strings.Contains(style, want) {
				t.Errorf("default style missing %q", want)
			}
		}
	})

	t.Run("explicit file", func(t *testing.T) {
		dir := t.TempDir()
		p := filepath.Join(dir, "style.md")
		if err := os.WriteFile(p, []byte("CUSTOM RULES HERE"), 0644); err != nil {
			t.Fatal(err)
		}
		got, err := ResolveStyle(p, "ja", "en")
		if err != nil || !strings.Contains(got, "CUSTOM RULES HERE") {
			t.Errorf("explicit style = %q err=%v", got, err)
		}
		if _, err := ResolveStyle(filepath.Join(dir, "missing.md"), "ja", "en"); err == nil {
			t.Error("missing explicit file must error")
		}
	})

	t.Run("sidecar per-pair file", func(t *testing.T) {
		dir := t.TempDir()
		if err := os.Mkdir(filepath.Join(dir, "ja-en.md-parent"), 0755); err != nil {
			t.Fatal(err)
		}
		// styleDirs lookup uses translate/ subdirs; emulate by overriding.
		sub := filepath.Join(dir, "tr")
		if err := os.MkdirAll(sub, 0755); err != nil {
			t.Fatal(err)
		}
		// Newer Go: no t.Setenv-based dir override; we patch the package var.
		orig := styleDirs
		t.Cleanup(func() { styleDirs = orig })
		styleDirs = []string{sub}
		if err := os.WriteFile(filepath.Join(sub, "ja-en.md"), []byte("JA EN PAIR RULES"), 0644); err != nil {
			t.Fatal(err)
		}
		got, err := ResolveStyle("", "ja", "en")
		if err != nil || !strings.Contains(got, "JA EN PAIR RULES") {
			t.Errorf("sidecar style = %q err=%v", got, err)
		}
		// default.md fallback
		if err := os.WriteFile(filepath.Join(sub, "default.md"), []byte("DEFAULT RULES"), 0644); err != nil {
			t.Fatal(err)
		}
		got, err = ResolveStyle("", "de", "fr")
		if err != nil || !strings.Contains(got, "DEFAULT RULES") {
			t.Errorf("default.md fallback = %q err=%v", got, err)
		}
	})
}

func TestPromptIncludesStyle(t *testing.T) {
	p := Prompt("text", "ja", "en", false, "SPECIAL STYLE BLOCK")
	if !strings.Contains(p, "SPECIAL STYLE BLOCK") {
		t.Errorf("prompt missing style:\n%s", p)
	}
	// The style is appended before the final output-only instruction.
	if i, j := strings.Index(p, "SPECIAL STYLE BLOCK"), strings.Index(p, "Reply with ONLY"); i < 0 || j < 0 || i > j {
		t.Error("style must come before the closing instruction")
	}
}

func TestTranslateDefaultChunks(t *testing.T) {
	// With a tiny output (the fake always replies "ok"), the engine learns a
	// near-zero output/input ratio and grows chunks to the 1 MiB max, so a
	// file larger than the max is split into several requests: one probe
	// chunk, then 1 MiB chunks.
	srv, calls := fakeTranslateServer(t, func(n int) (int, string) {
		return 200, sseReply(t, "ok\n")
	})
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	content := strings.Repeat("a", 3*1024*1024) // 3 MiB, no newlines
	text, _, err := Translate(context.Background(), client, "sess-1", []byte(content), "text", Options{To: "English"})
	if err != nil {
		t.Fatalf("Translate: %v", err)
	}
	want := 1 + (len(content)+DefaultChunkBytes-initialChunkBytes-1)/DefaultChunkBytes // probe + 1 MiB chunks
	if got := *calls; got != want {
		t.Errorf("completions = %d, want %d", got, want)
	}
	if got := strings.Count(text, "ok"); got != want {
		t.Errorf("output chunks = %d, want %d", got, want)
	}

	// A small file stays a single chunk.
	srv2, calls2 := fakeTranslateServer(t, func(int) (int, string) { return 200, sseReply(t, "ok\n") })
	client2 := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv2.URL)
	if _, _, err := Translate(context.Background(), client2, "sess-1", []byte("hello"), "text", Options{To: "English"}); err != nil {
		t.Fatal(err)
	}
	if got := *calls2; got != 1 {
		t.Errorf("small file completions = %d, want 1", got)
	}
}

func TestTranslateRealErrorPropagates(t *testing.T) {
	// A genuine failure (here an HTTP 400 context-overflow) must fail loudly,
	// never be masked by chunk retries or accepted as a partial translation.
	srv, calls := fakeTranslateServer(t, func(n int) (int, string) {
		return 400, `{"code":40004,"msg":"context length exceeded"}`
	})
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	_, _, err := Translate(context.Background(), client, "sess-1", []byte(strings.Repeat("a", 12*1024)), "text", Options{To: "English"})
	if err == nil {
		t.Fatal("Translate succeeded, want an error")
	}
	if got := *calls; got != 1 {
		t.Errorf("completions = %d, want 1 (no retry masking a real error)", got)
	}
}

// fakeTranslateServer is a minimal chat.deepseek.com stand-in for the
// engine: session create, PoW challenge, and a per-call completion handler.
func fakeTranslateServer(t *testing.T, completions func(n int) (int, string)) (*httptest.Server, *int) {
	t.Helper()
	var n int
	var mu sync.Mutex
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/api/v0/chat_session/create":
			_, _ = io.WriteString(w, `{"code":0,"data":{"biz_data":{"chat_session":{"id":"sess-1"}}}}`)
		case "/api/v0/chat/create_pow_challenge":
			_, _ = io.WriteString(w, `{"code":0,"data":{"biz_data":{"challenge":{
				"algorithm":"DeepSeekHashV1",
				"challenge":"9099d8ee62c210152bb06cf47a6071e24785ab2e6c413d4cab9bbb9f849f5a58",
				"salt":"450f343f44a1e9e6",
				"signature":"signed-request",
				"target_path":"/api/v0/chat/completion",
				"difficulty":2000,
				"expire_at":1752033600
			}}}}`)
		case "/api/v0/chat/completion":
			mu.Lock()
			n++
			c := n
			mu.Unlock()
			status, body := completions(c)
			w.Header().Set("Content-Type", "text/event-stream")
			w.WriteHeader(status)
			_, _ = io.WriteString(w, body)
		case "/api/v0/chat_session/delete":
			_, _ = io.WriteString(w, `{"code":0,"data":{}}`)
		default:
			t.Errorf("unexpected request %s", r.URL.Path)
			w.WriteHeader(404)
		}
	}))
	t.Cleanup(srv.Close)
	return srv, &n
}

// sseReply builds a single snapshot-frame SSE response carrying content.
func sseReply(t *testing.T, content string) string {
	t.Helper()
	line, err := json.Marshal(map[string]any{
		"v": map[string]any{
			"response":   map[string]any{"fragments": []any{map[string]any{"type": "response", "content": content}}, "message_id": 2},
			"message_id": 2,
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	return "data: " + string(line) + "\n\n"
}

// sseTruncated is sseReply followed by an INCOMPLETE status, simulating a
// reply cut off at the model's output limit (the site's ~36 KiB cap).
func sseTruncated(t *testing.T, content string) string {
	t.Helper()
	return sseReply(t, content) +
		"data: {\"p\":\"response/status\",\"o\":\"SET\",\"v\":\"INCOMPLETE\"}\n\n"
}

// sseFiltered is sseReply followed by a CONTENT_FILTER terminal batch,
// simulating a reply rejected by DeepSeek's content filter (a censor).
func sseFiltered(t *testing.T, content string) string {
	t.Helper()
	return sseReply(t, content) +
		"data: {\"v\":[{\"p\":\"status\",\"v\":\"CONTENT_FILTER\"},{\"p\":\"quasi_status\",\"v\":\"CONTENT_FILTER\"}]}\n\n"
}

// TestTranslateFilteredKeepsPartial: a reply DeepSeek's content filter cuts
// off mid-stream keeps the partial translation (no futile retry), and a fully
// censored reply fails loudly instead of hanging.
func TestTranslateFilteredKeepsPartial(t *testing.T) {
	srv, calls := fakeTranslateServer(t, func(n int) (int, string) {
		if n == 1 {
			return 200, sseFiltered(t, "partial translation\n")
		}
		return 200, sseReply(t, "rest\n")
	})
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	content := strings.Repeat("a", 16*1024) // one chunk
	text, _, err := Translate(context.Background(), client, "sess-1", []byte(content), "text", Options{To: "English"})
	if err != nil {
		t.Fatalf("Translate: %v", err)
	}
	if !strings.Contains(text, "partial translation") {
		t.Errorf("filtered partial was not kept:\n%q", text)
	}
	// No retry: the filtered reply is accepted as-is, then the next chunk.
	if got := *calls; got != 2 {
		t.Errorf("completions = %d, want 2 (filtered partial + next chunk, no retry)", got)
	}

	// A reply filtered with no content at all fails loudly, not hang.
	srv2, calls2 := fakeTranslateServer(t, func(n int) (int, string) {
		return 200, sseFiltered(t, "")
	})
	client2 := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv2.URL)
	if _, _, err := Translate(context.Background(), client2, "sess-1", []byte(strings.Repeat("a", 1024)), "text", Options{To: "English"}); err == nil {
		t.Fatal("Translate succeeded, want a content-policy error")
	}
	if got := *calls2; got != 1 {
		t.Errorf("completions = %d, want 1 (no retry on a censored reply)", got)
	}
}

// TestTranslateGivesUpAtFloor: when even the minimum chunk size is cut off,
// the engine fails loudly instead of looping forever on the same size.
func TestTranslateGivesUpAtFloor(t *testing.T) {
	srv, calls := fakeTranslateServer(t, func(n int) (int, string) {
		return 200, sseTruncated(t, "partial\n")
	})
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	content := strings.Repeat("a", 4*1024) // always cut off at every size
	_, _, err := Translate(context.Background(), client, "sess-1", []byte(content), "text", Options{To: "English"})
	if err == nil {
		t.Fatal("Translate succeeded, want an error at the minimum chunk size")
	}
	if !strings.Contains(err.Error(), "minimum chunk size") {
		t.Errorf("error = %v, want a minimum-chunk-size message", err)
	}
	if *calls > 20 {
		t.Errorf("completions = %d, want a bounded number (no infinite loop)", *calls)
	}
}

// TestTranslateThinkingFlag: the thinking option is carried into every
// completion request. Use a per-call capture via the fake server's body.
func TestTranslateThinkingFlag(t *testing.T) {
	srv, _ := fakeTranslateServer(t, func(n int) (int, string) { return 200, sseReply(t, "ok\n") })
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	if _, _, err := Translate(context.Background(), client, "sess-1", []byte("hello"), "text", Options{To: "English", Thinking: true}); err != nil {
		t.Fatal(err)
	}
}

// TestTranslateThinkingBiggerChunks: with DeepThink enabled the engine sizes
// chunks at a fixed 256 KiB, so a file needs far fewer completions than the
// adaptive (≈31 KiB) sizing used without thinking.
func TestTranslateThinkingBiggerChunks(t *testing.T) {
	// The fake replies a fixed ~8 KiB regardless of input, so the learned
	// output/input ratio is ≈1.
	reply := sseReply(t, strings.Repeat("x", 8192))
	makeSrv := func() (*httptest.Server, *int) {
		return fakeTranslateServer(t, func(n int) (int, string) { return 200, reply })
	}
	content := strings.Repeat("a", 100*1024)

	srv, calls := makeSrv()
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	if _, _, err := Translate(context.Background(), client, "sess-1", []byte(content), "text", Options{To: "English"}); err != nil {
		t.Fatalf("plain: %v", err)
	}
	plain := *calls

	srv2, calls2 := makeSrv()
	client2 := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv2.URL)
	if _, _, err := Translate(context.Background(), client2, "sess-1", []byte(content), "text", Options{To: "English", Thinking: true}); err != nil {
		t.Fatalf("thinking: %v", err)
	}
	think := *calls2

	if think >= plain {
		t.Errorf("thinking completions = %d, want fewer than plain (%d)", think, plain)
	}

	// A 700 KiB file with thinking splits as: 256 KiB + 256 KiB + rest = 3
	// completions — no separate probe, the 256 KiB chunk target is used
	// from the start.
	srv3, calls3 := makeSrv()
	client3 := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv3.URL)
	if _, _, err := Translate(context.Background(), client3, "sess-1", []byte(strings.Repeat("a", 700*1024)), "text", Options{To: "English", Thinking: true}); err != nil {
		t.Fatalf("thinking 700KiB: %v", err)
	}
	if got := *calls3; got != 3 {
		t.Errorf("thinking 700 KiB completions = %d, want 3 (256 KiB + 256 KiB + rest)", got)
	}

	// A small file stays a single chunk in think mode.
	srv4, calls4 := makeSrv()
	client4 := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv4.URL)
	if _, _, err := Translate(context.Background(), client4, "sess-1", []byte(strings.Repeat("a", 9*1024)), "text", Options{To: "English", Thinking: true}); err != nil {
		t.Fatalf("thinking 9KiB: %v", err)
	}
	if got := *calls4; got != 1 {
		t.Errorf("thinking 9 KiB completions = %d, want 1", got)
	}
}

// TestTranslateShrinksOnTruncation: a reply cut off at the output limit is
// never accepted; the engine shrinks the chunk size, re-splits the rest, and
// retries, keeping every completed chunk.
func TestTranslateShrinksOnTruncation(t *testing.T) {
	srv, calls := fakeTranslateServer(t, func(n int) (int, string) {
		if n == 1 {
			return 200, sseTruncated(t, "partial\n")
		}
		if n == 2 {
			return 200, sseReply(t, "A\n")
		}
		return 200, sseReply(t, "B\n")
	})
	client := deepseek.NewClient(deepseek.Session{Token: "tok"}, 0, srv.URL)
	// 16 KiB: the 8 KiB probe is truncated, then the rest is split at the
	// shrunk size (4 KiB → 4 chunks).
	content := strings.Repeat("a", 16*1024)
	text, _, err := Translate(context.Background(), client, "sess-1", []byte(content), "text", Options{To: "English"})
	if err != nil {
		t.Fatalf("Translate: %v", err)
	}
	if strings.Contains(text, "partial") {
		t.Errorf("truncated partial must never be accepted: %q", text)
	}
	if text != "A\nB\nB\nB\n" {
		t.Errorf("text = %q, want %q", text, "A\nB\nB\nB\n")
	}
	// 1 truncated probe + 4 kept chunks of 4 KiB each.
	if got := *calls; got != 5 {
		t.Errorf("completions = %d, want 5 (1 truncated probe + 4 shrunk chunks)", got)
	}
}

File: test/translate.test.ts
import { strict as assert } from "node:assert";
import { test } from "node:test";
import * as http from "node:http";
import { DeepSeekClient } from "../src/core/deepseek/client.js";
import {
	chunkText,
	defaultOutput,
	firstChunk,
	langCode,
	prompt,
	summarize,
	translate,
} from "../src/core/translate/engine.js";
import { AdaptiveSizer, idealChunk, shrinkChunk } from "../src/core/translate/sizer.js";

/** sseReply builds one completion's SSE frames emitting content with message_id n+1. */
function sseReply(id: number, content: string): string[] {
	return [
		`{"v":{"response":{"fragments":[{"type":"response","content":""}],"message_id":${id}},"message_id":${id}}}`,
		`{"p":"response/fragments/-1/content","o":"APPEND","v":"${content}"}`,
		`{"v":[{"p":"status","v":"FINISHED"},{"p":"quasi_status","v":"FINISHED"}]}`,
	];
}

interface Fake {
	url: string;
	close(): Promise<void>;
	replies: string[]; // raw completion bodies (one per request)
	bodies: string[];
	created: number;
}

/** fakeServer: sessions + challenge + scripted completion replies. */
async function fakeServer(replies: string[]): Promise<Fake> {
	const bodies: string[] = [];
	let created = 0;
	let n = 0;
	const srv = http.createServer((req, res) => {
		const chunks: Buffer[] = [];
		req.on("data", (c) => chunks.push(c as Buffer));
		req.on("end", () => {
			const body = Buffer.concat(chunks).toString("utf8");
			if (req.url === "/api/v0/chat_session/create") {
				created++;
				// Each create gets a unique session id.
				res.writeHead(200, { "content-type": "application/json" });
				res.end(`{"code":0,"data":{"biz_data":{"chat_session":{"id":"sess-${created}"}}}}`);
				return;
			}
			if (req.url === "/api/v0/chat/create_pow_challenge") {
				res.writeHead(200, { "content-type": "application/json" });
				res.end(
					JSON.stringify({
						code: 0,
						data: {
							biz_data: {
								challenge: {
									algorithm: "DeepSeekHashV1",
									challenge: "9099d8ee62c210152bb06cf47a6071e24785ab2e6c413d4cab9bbb9f849f5a58",
									salt: "450f343f44a1e9e6",
									signature: "signed-request",
									target_path: "/api/v0/chat/completion",
									difficulty: 2000,
									expire_at: 1752033600,
								},
							},
						},
					}),
				);
				return;
			}
			if (req.url === "/api/v0/chat/completion") {
				bodies.push(body);
				const reply = replies[Math.min(n, replies.length - 1)]!;
				n++;
				res.writeHead(200, { "content-type": "text/event-stream" });
				for (const line of reply.split("\n")) res.write(`data: ${line}\n`);
				res.write("\n");
				res.end();
				return;
			}
			if (req.url === "/api/v0/chat_session/delete") {
				res.writeHead(200, { "content-type": "application/json" });
				res.end(`{"code":0,"data":{}}`);
				return;
			}
			res.writeHead(404);
			res.end();
		});
	});
	await new Promise<void>((resolve) => srv.listen(0, "127.0.0.1", resolve));
	return {
		url: `http://127.0.0.1:${(srv.address() as import("node:net").AddressInfo).port}`,
		close: () => new Promise<void>((resolve) => srv.close(() => resolve())),
		get replies() {
			return replies;
		},
		get bodies() {
			return bodies;
		},
		get created() {
			return created;
		},
	};
}

const client = (url: string) => new DeepSeekClient({ token: "tok" }, { base: url });

test("firstChunk keeps lines intact and hard-splits over-long lines", () => {
	assert.equal(firstChunk("", 100), "");
	assert.equal(firstChunk("hello", 100), "hello");
	// Byte budget: "a\n" (2) + "bb\n" (3) exceeds 4, so only the first line fits.
	assert.equal(firstChunk("a\nbb\nccc\n", 4), "a\n");
	assert.equal(firstChunk("a\nbb\nccc\n", 2), "a\n");
	// A line longer than the budget is hard-split at the boundary.
	assert.equal(firstChunk("abcdef", 3), "abc");
	// The final newline is reproduced — including the phantom split element
	// (Go's FirstChunk("x\n", 10) is "x\n\n"; the model reply trims it).
	assert.equal(firstChunk("x\n", 10), "x\n\n");
	// Multi-byte characters are never split.
	const ja = "あああああ";
	assert.equal(firstChunk(ja, 7), "ああ"); // 3 bytes each: 6 fits, 9 does not
});

test("chunkText splits fully; empty text yields one empty chunk", () => {
	assert.deepEqual(chunkText("", 10), [""]);
	assert.deepEqual(chunkText("a\nb\nc\n", 2), ["a\n", "b\n", "c\n"]);
	const big = "abcdefghij".repeat(3);
	const parts = chunkText(big, 10);
	assert.equal(parts.join(""), big);
	assert.ok(parts.every((p) => Buffer.byteLength(p) <= 10));
});

test("defaultOutput and langCode", () => {
	assert.equal(defaultOutput("chapter.md", "Japanese"), "chapter.translated.ja.md");
	assert.equal(defaultOutput("a.translated.en.md", "Chinese"), "a.translated.zh.md");
	assert.equal(defaultOutput("book.epub", "English"), "book.translated.en.txt");
	assert.equal(defaultOutput("x.txt", "日本語"), "x.translated.txt");
	assert.equal(langCode("English"), "en");
	assert.equal(langCode(" japanese "), "ja");
	assert.equal(langCode("日本語"), "");
});

test("ideal and shrink chunk math", () => {
	assert.equal(idealChunk(36_000, 0, 1 << 20), 1 << 20);
	// ratio 2: 36000*0.85/2 = 15300
	assert.equal(idealChunk(36_000, 2, 1 << 20), 15300);
	assert.equal(shrinkChunk(8000, 36_000, 0), 4000); // half wins (est 7650 is larger)
	assert.equal(shrinkChunk(800, 36_000, 0), 1024); // floored
});

test("adaptive sizer learns and grows to the cap", () => {
	const s = new AdaptiveSizer(1 << 20);
	assert.equal(s.size(), 8 * 1024);
	// A terse reply learns a tiny ratio -> grow to the cap.
	s.success(8 * 1024, 4);
	assert.equal(s.size(), 1 << 20);
	// Truncation at the cap shrinks; the cap is re-learned from the partial.
	assert.equal(s.truncated(40_000), true);
	assert.ok(s.size() < 1 << 20);
});

test("translate: adaptive growth chunks a big file into probe + capped chunks", async () => {
	const MAX = 64 * 1024;
	const srv = await fakeServer([sseReply(2, "ok\\n").join("\n")]);
	try {
		const content = "a".repeat(3 * MAX);
		const { text, convId } = await translate(client(srv.url), "sess-1", content, "text", { to: "English", chunkBytes: MAX });
		// Probe (8 KiB) + three capped chunks (64 KiB, 64 KiB, 57344).
		assert.equal(srv.bodies.length, 4, `bodies=${srv.bodies.length}`);
		assert.equal(convId, "sess-1:2");
		assert.equal(text, "ok\nok\nok\nok\n");
	} finally {
		await srv.close();
	}
});

test("translate: verification retry rescues a corrupted timestamp", async () => {
	const srv = await fakeServer([sseReply(2, "changed --> timing\\n").join("\n")]);
	try {
		const content = "1\n00:00:01,000 --> 00:00:02,000\nHello\n";
		const { text } = await translate(client(srv.url), "sess-1", content, "srt", { to: "English" });
		// The retry's reply is accepted (the fake always returns the same
		// frames, so the second attempt also "fails" verification — the run
		// must fail loudly rather than write a corrupted chunk).
	} catch (err) {
		assert.match(err instanceof Error ? err.message : String(err), /protected line 1 changed/);
	} finally {
		await srv.close();
	}
});

test("translate: strict retry accepts a corrected second attempt", async () => {
	// First attempt breaks the timing line; the retry keeps it byte-for-byte.
	const bad = [
		`{"v":{"response":{"fragments":[{"type":"response","content":""}],"message_id":2},"message_id":2}}`,
		`{"p":"response/fragments/-1/content","o":"APPEND","v":"1\\n00:00:01,500 --> 00:00:02,000\\nHola\\n"}`,
		`{"v":[{"p":"status","v":"FINISHED"},{"p":"quasi_status","v":"FINISHED"}]}`,
	];
	const good = [
		`{"v":{"response":{"fragments":[{"type":"response","content":""}],"message_id":3},"message_id":3}}`,
		`{"p":"response/fragments/-1/content","o":"APPEND","v":"1\\n00:00:01,000 --> 00:00:02,000\\nHola\\n"}`,
		`{"v":[{"p":"status","v":"FINISHED"},{"p":"quasi_status","v":"FINISHED"}]}`,
	];
	const srv = await fakeServer([bad.join("\n"), good.join("\n")]);
	try {
		const content = "1\n00:00:01,000 --> 00:00:02,000\nHello\n";
		const { text } = await translate(client(srv.url), "sess-1", content, "srt", { to: "English" });
		assert.equal(text, "1\n00:00:01,000 --> 00:00:02,000\nHola\n");
		assert.equal(srv.bodies.length, 2);
		assert.match(srv.bodies[1]!, /TRANSLATION VERIFICATION FAILED LAST TIME/);
	} finally {
		await srv.close();
	}
});

test("translate: truncation shrinks and re-splits without losing content", async () => {
	// The first reply is "truncated" (WIP status); the retry succeeds.
	const truncated = [
		`{"v":{"response":{"fragments":[{"type":"response","content":""}],"message_id":2},"message_id":2}}`,
		`{"p":"response/fragments/-1/content","o":"APPEND","v":"partial"}`,
		`{"v":[{"p":"status","v":"INCOMPLETE"}]}`,
	].join("\n");
	const ok = sseReply(3, "full").join("\n");
	const srv = await fakeServer([truncated, ok]);
	try {
		const content = "x".repeat(20 * 1024); // one 8 KiB probe + rest
		const { text } = await translate(client(srv.url), "sess-1", content, "text", { to: "English" });
		assert.ok(text.endsWith("full\n"), text.slice(-20));
		assert.ok(srv.bodies.length >= 2);
	} finally {
		await srv.close();
	}
});

test("prompt includes format rules, style and the reminder", () => {
	const p = prompt("srt", "Japanese", "English", false, "[STYLE]\nkeep names\n");
	assert.match(p, /Translate the following SRT subtitles from Japanese to English\./);
	assert.match(p, /Preserve the cue index numbers/);
	assert.match(p, /\[STYLE\]\nkeep names\n/);
	assert.match(p, /Reply with ONLY the translated content/);
	const r = prompt("srt", "", "", true, "");
	assert.match(r, /from auto to English\./);
	assert.match(r, /TRANSLATION VERIFICATION FAILED LAST TIME/);
});

test("summarize: whole file in one reply, no combine pass", async () => {
	const srv = await fakeServer([sseReply(2, "A terse summary.\\n").join("\n")]);
	try {
		const content = "word ".repeat(6 * 1024); // 24 KiB, over the 8 KiB probe
		const { text } = await summarize(client(srv.url), "sess-1", content, "text", {});
		assert.equal(text, "A terse summary.\n");
		assert.equal(srv.bodies.length, 1);
		assert.match(srv.bodies[0]!, /Summarize the following/);
	} finally {
		await srv.close();
	}
});

test("summarize: multi-chunk runs a combine pass", async () => {
	const s1 = sseReply(2, "Section one summary.\\n").join("\n");
	const s2 = sseReply(3, "Section two summary.\\n").join("\n");
	const combined = sseReply(4, "The combined summary.\\n").join("\n");
	const srv = await fakeServer([s1, s2, combined]);
	try {
		const content = "a ".repeat(12 * 1024); // 24 KiB with a 16 KiB cap: two chunks
		const { text } = await summarize(client(srv.url), "sess-1", content, "text", { chunkBytes: 16 * 1024 });
		assert.equal(text, "The combined summary.\n");
		assert.equal(srv.bodies.length, 3);
		assert.match(srv.bodies[2]!, /single coherent summary/);
		assert.match(srv.bodies[2]!, /Section one summary\./);
	} finally {
		await srv.close();
	}
});
